storage/client_test.go

```diff
@@ -1506,6 +1506,226 @@ func TestRetryReadStallEmulated(t *testing.T) {
+					Generation: defaultGen,
+				}
+
+				var gotAttrs *ObjectAttrs
+				params := &openWriterParams{
+					attrs:     want,
+					bucket:    bucket,
+					chunkSize: chunkSize,
+					ctx:       ctx,
+					donec:     make(chan struct{}),
+					setError:  func(_ error) {}, // no-op
+					progress:  func(_ int64) {}, // no-op
+					setObj:    func(o *ObjectAttrs) { gotAttrs = o },
+				}
+
+				startTime := time.Now()
+				pw, err := client.OpenWriter(params)
+				if err != nil {
+					t.Fatalf("failed to open writer: %v", err)
+				}
+				buffer := bytes.Repeat([]byte("A"), fileSize)
+				if _, err := pw.Write(buffer); err != nil {
+					t.Fatalf("failed to populate test data: %v", err)
+				}
+				if err := pw.Close(); err != nil {
+					t.Fatalf("closing object: %v", err)
+				}
+				endTime := time.Now()
+				select {
+				case <-params.donec:
+				}
+				if gotAttrs == nil {
+					t.Fatalf("Writer finished, but resulting object wasn't set")
+				}
+				if diff := cmp.Diff(gotAttrs.Name, want.Name); diff != "" {
+					t.Fatalf("Resulting object name: got(-),want(+):\n%s", diff)
+				}
+				if endTime.Sub(startTime) < tc.wantDuration {
+					t.Errorf("write took %v, want >= %v", endTime.Sub(startTime), tc.wantDuration)
+				}
+
+				r, err := veneerClient.Bucket(bucket).Object(want.Name).NewReader(ctx)
+				if err != nil {
+					t.Fatalf("opening reading: %v", err)
+				}
+				wantLen := len(buffer)
+				got := make([]byte, wantLen)
+				n, err := r.Read(got)
+				if n != wantLen {
+					t.Fatalf("expected to read %d bytes, but got %d", wantLen, n)
+				}
+				if diff := cmp.Diff(got, buffer); diff != "" {
+					t.Fatalf("checking written content: got(-),want(+):\n%s", diff)
+				}
+			})
+		}
+	})
+}
+
+// In resumable uploads, the chunk encounter a stall. The chunkTransferTimeout will then cancel the request and resend it.
+// The second request is expected to succeed. This allows the upload to finish earlier than if it had waited for the entire stall
+// duration, as the chunkTransferTimeout prevents unnecessary delays.
+func TestRetryWriteReqStallWithNonZeroChunkTransferTimeoutEmulated(t *testing.T) {
+	transportClientTest(skipGRPC("service is not implemented"), t, func(t *testing.T, ctx context.Context, project, bucket string, client storageClient) {
+		_, err := client.CreateBucket(ctx, project, bucket, &BucketAttrs{}, nil)
+		if err != nil {
+			t.Fatalf("creating bucket: %v", err)
+		}
+
+		chunkSize := 2 * 1024 * 1024 // 2 MiB
+		fileSize := 5 * 1024 * 1024  // 5 MiB
+		tests := []struct {
+			name                 string
+			instructions         map[string][]string
+			chunkTransferTimeout time.Duration
+			wantDuration         time.Duration
+		}{
+			{
+				name: "stall-on-first-chunk-with-chunk-transfer-timeout-nonzero",
+				instructions: map[string][]string{
+					"storage.objects.insert": {"stall-for-10s-after-1024K"},
+				},
+				chunkTransferTimeout: 2 * time.Second,
```

Comment:
I would set the chunk transfer timeout shorter, maybe 100ms. We are working with a local emulator so latency should be lower than for the real service.